kernel/kernel-api/src/main/java/io/delta/kernel/internal/actions/AddFile.java

```diff
@@ -55,12 +60,15 @@ public class AddFile {
           .add(
               "tags",
               new MapType(StringType.STRING, StringType.STRING, true /* valueContainsNull */),
-              true /* nullable */);
+              true /* nullable */)
+          .add("baseRowId", LongType.LONG, true /* nullable */)
+          .add("defaultRowCommitVersion", LongType.LONG, true /* nullable */);
 
   public static final StructType SCHEMA_WITH_STATS = SCHEMA_WITHOUT_STATS.add(JSON_STATS_FIELD);
 
   /** Full schema of the {@code add} action in the Delta Log. */
   public static final StructType FULL_SCHEMA = SCHEMA_WITH_STATS;
+
   // There are more fields which are added when row-id tracking and clustering is enabled.
   // When Kernel starts supporting row-ids and clustering, we should add those fields here.
 
```

Comment:
If we have a way to access `COL_NAME_TO_ORDINAL` from `RowBackedAction`, we should be able to move the common logic to access and update fields there and make it easier to implement child classes without as much repetition.

We can make it a non-static member variable of  the action though as that would take up space for every instance. We can't make it static either, as it depends on the concrete action schema.

We could stop pre-computing it and just implement this as a method `getColumnIndex()` that looks up the name in the schema. That's less efficient though.. unless we keep using a map to quickly lookup fields, which `StructType` already contains internally:
https://github.com/delta-io/delta/blob/5a39f7047e37d05d61aef39ba93a5c6b1b3ebd6c/kernel/kernel-api/src/main/java/io/delta/kernel/types/StructType.java#L32
```
  private final Map<String, Tuple2<StructField, Integer>> nameToFieldAndOrdinal;
```


So the following should be as efficient as today:
```
protected Integer indexOf(String fieldName) {
  return getFullSchema().indexOf(fieldName);
}
```
and implement indexOf as a lookup in `nameToFieldAndOrdinal` instead of an array lookup as it's done currently.

You can then have in `RowBackedAction`, e.g.:

```
protected String getString(String fieldName) {
  return row.getString(indexOf(fieldName));
}
```
```
  public String getPath() {
    return getString("path");
  }
```

Or:
```
  protected Row withOverridenValue(String fieldName, Object newValue) {
    Map<Integer, Object> overrides =
        Collections.singletonMap(indexOf(fieldName), newValue);
    return new AddFile(new DelegateRow(row, overrides));
  }
```
```
  public AddFile withNewBaseRowId(long baseRowId) {
    return new AddFile(withOverridenValue("baseRowId", baseRowId));
  }
```

With this, you should actually be able to get to a point where `row` can be private and is only accessed through the base class

Reply:
Thanks for the suggestion! I’ve implemented the changes. `row` is now private to `RowBackedAction`.